In [1]:
import pandas as pd
import altair as alt

In [2]:
#allows altair to use unlimited number of rows when visualizing data
alt.data_transformers.disable_max_rows()

DataTransformerRegistry.enable('default')

In [3]:
roster = pd.read_csv("/Users/josemartinez/Desktop/Code/Discipline_Data/staff_roster_cleaned.csv")

In [4]:
roster.columns

Index(['name', 'job_desc', 'dept_desc', 'gender', 'race_ethnicity',
       'date_started', 'year', 'clean_name', 'clean_race_ethnicity'],
      dtype='object')

In [5]:
#Pulled out race and year from dataset
year_race = roster[['year','clean_race_ethnicity']]

# **Demographic breakdown by year**

### Demographic Breakdown in 2010

In [6]:
#filtered data by every row that had 2010 as the roster year
roster_2010 = year_race.loc[year_race['year'] == 2010]
#counted number of instances that each value had in new data (ex. 50 hispanics 50 blacks)
chart10 = roster_2010.value_counts().reset_index()
#don't need year column anymore since it's all 2020, so I deleted it
del chart10['year']
#renamed columns for clarity
chart10.columns = ['Race', 'Amount']

In [7]:
#Made a new column that found the percentage of each race group and rounded it to one decimanl point
chart10['Percentage'] = ((chart10['Amount']/chart10['Amount'].sum())*100).round(1)
#Added the string '%' at the end for chart marking purposes.
chart10['Percentage'] = chart10['Percentage'].astype(str) + '%'

In [8]:
chart10.style.set_caption("Officer Demographics MNPD - 2010")

In [9]:
ten = alt.Chart(chart10, title= 'Officer Demographics MNPD - 2010').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

#created the percentage labels to put above bars on chart
#'dy' just moves the label vertically while the 'text' command chooses what column to pull from the database
text = ten.mark_text(dy=-10).encode(text='Percentage')

#Combine the chart and text labels
ten + text

alt.LayerChart(...)

# Notes above are the same for the following yearly demographic breakdowns

### Demographic Breakdown in 2011

In [10]:
roster_2011 = year_race.loc[year_race['year'] == 2011]
chart11 = roster_2011.value_counts().reset_index()
del chart11['year']
chart11.columns = ['Race', 'Amount']

In [11]:
chart11['Percentage'] = ((chart11['Amount']/chart11['Amount'].sum())*100).round(1)
chart11['Percentage'] = chart11['Percentage'].astype(str) + '%'

In [12]:
chart11

Race  Amount Percentage
0                   white    1379      75.9%
1                   black     388      21.4%
2                hispanic      24       1.3%
3  asian/pacific islander      22       1.2%
4         native american       2       0.1%
5                 unknown       1       0.1%

In [13]:
eleven = alt.Chart(chart11, title= 'Officer Demographics MNPD - 2011').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = eleven.mark_text(dy=-10).encode(text='Percentage')

eleven + text

alt.LayerChart(...)

### Demographic Breakdown in 2012

In [14]:
roster_2012 = year_race.loc[year_race['year'] == 2012]
chart12 = roster_2012.value_counts().reset_index()
del chart12['year']
chart12.columns = ['Race', 'Amount']

In [15]:
chart12['Percentage'] = ((chart12['Amount']/chart12['Amount'].sum())*100).round(1)
chart12['Percentage'] = chart12['Percentage'].astype(str) + '%'

In [16]:
chart12

Race  Amount Percentage
0                   white    1547      76.2%
1                   black     429      21.1%
2                hispanic      26       1.3%
3  asian/pacific islander      23       1.1%
4         native american       3       0.1%
5                 unknown       1       0.0%

In [17]:
twelve = alt.Chart(chart12, title= 'Officer Demographics MNPD - 2012').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = twelve.mark_text(dy=-10).encode(text='Percentage')

twelve + text

alt.LayerChart(...)

### Demographic Breakdown in 2013

In [18]:
roster_2013 = year_race.loc[year_race['year'] == 2013]
chart13 = roster_2013.value_counts().reset_index()
del chart13['year']
chart13.columns = ['Race', 'Amount']

In [19]:
chart13['Percentage'] = ((chart13['Amount']/chart13['Amount'].sum())*100).round(1)
chart13['Percentage'] = chart13['Percentage'].astype(str) + '%'

In [20]:
chart13

Race  Amount Percentage
0                   white    1493      77.2%
1                   black     381      19.7%
2                hispanic      30       1.6%
3  asian/pacific islander      25       1.3%
4         native american       2       0.1%
5                 unknown       2       0.1%

In [21]:
thirteen = alt.Chart(chart13, title= 'Officer Demographics MNPD - 2013').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = thirteen.mark_text(dy=-10).encode(text='Percentage')

thirteen + text

alt.LayerChart(...)

### Demographic Breakdown in 2014

In [22]:
roster_2014 = year_race.loc[year_race['year'] == 2014]
chart14 = roster_2014.value_counts().reset_index()
del chart14['year']
chart14.columns = ['Race', 'Amount']

In [23]:
chart14['Percentage'] = ((chart14['Amount']/chart14['Amount'].sum())*100).round(1)
chart14['Percentage'] = chart14['Percentage'].astype(str) + '%'

In [24]:
chart14

Race  Amount Percentage
0                   white    1517      78.6%
1                   black     357      18.5%
2                hispanic      28       1.5%
3  asian/pacific islander      24       1.2%
4         native american       2       0.1%
5                 unknown       2       0.1%

In [25]:
fourteen = alt.Chart(chart14, title= 'Officer Demographics MNPD - 2014').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = fourteen.mark_text(dy=-10).encode(text='Percentage')

fourteen + text

alt.LayerChart(...)

### Demographic Breakdown in 2015

In [26]:
roster_2015 = year_race.loc[year_race['year'] == 2015]
chart15 = roster_2015.value_counts().reset_index()
del chart15['year']
chart15.columns = ['Race', 'Amount']

In [27]:
chart15['Percentage'] = ((chart15['Amount']/chart15['Amount'].sum())*100).round(1)
chart15['Percentage'] = chart15['Percentage'].astype(str) + '%'

In [28]:
chart15

Race  Amount Percentage
0                   white    1552      79.4%
1                   black     342      17.5%
2                hispanic      31       1.6%
3  asian/pacific islander      25       1.3%
4         native american       2       0.1%
5                 unknown       2       0.1%

In [29]:
fifteen = alt.Chart(chart15, title= 'Officer Demographics MNPD - 2015').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = fifteen.mark_text(dy=-10).encode(text='Percentage')

fifteen + text

alt.LayerChart(...)

### Demographic Breakdown in 2016

In [30]:
roster_2016 = year_race.loc[year_race['year'] == 2016]
chart16 = roster_2016.value_counts().reset_index()
del chart16['year']
chart16.columns = ['Race', 'Amount']

In [31]:
chart16['Percentage'] = ((chart16['Amount']/chart16['Amount'].sum())*100).round(1)
chart16['Percentage'] = chart16['Percentage'].astype(str) + '%'

In [32]:
chart16

Race  Amount Percentage
0                   white    1587      78.8%
1                   black     361      17.9%
2                hispanic      36       1.8%
3  asian/pacific islander      26       1.3%
4                 unknown       3       0.1%
5         native american       1       0.0%

In [33]:
sixteen = alt.Chart(chart16, title= 'Officer Demographics MNPD - 2016').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = sixteen.mark_text(dy=-10).encode(text='Percentage')

sixteen + text

alt.LayerChart(...)

### Demographic Breakdown in 2017

In [34]:
roster_2017 = year_race.loc[year_race['year'] == 2017]
chart17 = roster_2017.value_counts().reset_index()
del chart17['year']
chart17.columns = ['Race', 'Amount']

In [35]:
chart17['Percentage'] = ((chart17['Amount']/chart17['Amount'].sum())*100).round(1)
chart17['Percentage'] = chart17['Percentage'].astype(str) + '%'

In [36]:
chart17

Race  Amount Percentage
0                   white    1586      78.4%
1                   black     373      18.4%
2                hispanic      36       1.8%
3  asian/pacific islander      25       1.2%
4                 unknown       3       0.1%
5         native american       1       0.0%

In [37]:
seventeen = alt.Chart(chart17, title= 'Officer Demographics MNPD - 2017').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = seventeen.mark_text(dy=-10).encode(text='Percentage')

seventeen + text

alt.LayerChart(...)

### Demographic Breakdown in 2018

In [38]:
roster_2018 = year_race.loc[year_race['year'] == 2018]
chart18 = roster_2018.value_counts().reset_index()
del chart18['year']
chart18.columns = ['Race', 'Amount']

In [39]:
chart18['Percentage'] = ((chart18['Amount']/chart18['Amount'].sum())*100).round(1)
chart18['Percentage'] = chart18['Percentage'].astype(str) + '%'

In [40]:
chart18

Race  Amount Percentage
0                   white    1602      76.1%
1                   black     369      17.5%
2             multiracial      68       3.2%
3                hispanic      41       1.9%
4  asian/pacific islander      21       1.0%
5         native american       4       0.2%

In [41]:
eighteen = alt.Chart(chart18, title= 'Officer Demographics MNPD - 2018').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = eighteen.mark_text(dy=-10).encode(text='Percentage')

eighteen + text

alt.LayerChart(...)

### Demographic Breakdown in 2019

In [42]:
roster_2019 = year_race.loc[year_race['year'] == 2019]
chart19 = roster_2019.value_counts().reset_index()
del chart19['year']
chart19.columns = ['Race', 'Amount']

In [43]:
chart19['Percentage'] = ((chart19['Amount']/chart19['Amount'].sum())*100).round(1)
chart19['Percentage'] = chart19['Percentage'].astype(str) + '%'

In [44]:
chart19

Race  Amount Percentage
0                   white    1508      76.4%
1                   black     332      16.8%
2             multiracial      64       3.2%
3                hispanic      45       2.3%
4  asian/pacific islander      22       1.1%
5         native american       3       0.2%

In [45]:
nineteen = alt.Chart(chart19, title= 'Officer Demographics MNPD - 2019').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = nineteen.mark_text(dy=-10).encode(text='Percentage')

nineteen + text

alt.LayerChart(...)

### Demographic Breakdown in 2020

In [46]:
roster_2020 = year_race.loc[year_race['year'] == 2020]
chart20 = roster_2020.value_counts().reset_index()
del chart20['year']
chart20.columns = ['Race', 'Amount']

In [47]:
chart20['Percentage'] = ((chart20['Amount']/chart20['Amount'].sum())*100).round(1)
chart20['Percentage'] = chart20['Percentage'].astype(str) + '%'

In [48]:
chart20

Race  Amount Percentage
0                   white    1506      76.2%
1                   black     330      16.7%
2             multiracial      61       3.1%
3                hispanic      52       2.6%
4  asian/pacific islander      24       1.2%
5         native american       4       0.2%

In [49]:
twenty = alt.Chart(chart20, title= 'Officer Demographics MNPD - 2020').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = twenty.mark_text(dy=-10).encode(text='Percentage')

twenty + text

alt.LayerChart(...)

### Demographic Breakdown in 2021

In [50]:
roster_2021 = year_race.loc[year_race['year'] == 2021]
chart21 = roster_2021.value_counts().reset_index()
del chart21['year']
chart21.columns = ['Race', 'Amount']

In [51]:
chart21['Percentage'] = ((chart21['Amount']/chart21['Amount'].sum())*100).round(1)
chart21['Percentage'] = chart21['Percentage'].astype(str) + '%'

In [52]:
chart21

Race  Amount Percentage
0                   white    1518      76.5%
1                   black     324      16.3%
2             multiracial      64       3.2%
3                hispanic      51       2.6%
4  asian/pacific islander      24       1.2%
5         native american       4       0.2%

In [53]:
twenty_one = alt.Chart(chart21, title= 'Officer Demographics MNPD - 2021').mark_bar().encode(
    alt.X('Race', axis=alt.Axis(labelAngle=0),sort=alt.EncodingSortField(field="Letters", op="count", order='ascending')),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

text = twenty_one.mark_text(dy=-10).encode(text='Percentage')

twenty_one + text

alt.LayerChart(...)

## Findings from yearly demographic breakdowns:
    • White officers's racial makeup at MNPD remained above 75% from 2010 to 2021.
    • Black officers were the next largest group. Their racial makeup decreased, starting at 20.9% in 2010 to 16.3% in 2021.
    • Multiracial was added as an option in 2018, so that could explain some percentage differences.
   

# **Unique Demographic Change over 11 years**

### White Officer Change from 2010 - 2021

In [54]:
#filtered dataset to where only white officers were shown over time
roster_white = year_race.loc[year_race['clean_race_ethnicity'] == 'white']
#found the number of white officers per year
chart_white = roster_white['year'].value_counts().reset_index()
#Renamed columns for clarity
chart_white.columns = ['Year','Amount']
#Turned the year column into string, because Altair takes in years as integers and starts chart at 0.
chart_white['Year'] = chart_white['Year'].astype(str)

In [55]:
chart_white

Year  Amount
0   2018    1602
1   2016    1587
2   2017    1586
3   2015    1552
4   2012    1547
5   2021    1518
6   2014    1517
7   2019    1508
8   2020    1506
9   2013    1493
10  2010    1455
11  2011    1379

In [56]:
chart_white['Amount'].mean()

1520.8333333333333

In [57]:
white_chart = alt.Chart(chart_white, title= 'White Officers from 2010 - 2021').mark_bar().encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

In [58]:
white_chart

alt.Chart(...)

## Notes above are same for following racial stats over time

### Black Officer Change from 2010 - 2021

In [59]:
roster_black = year_race.loc[year_race['clean_race_ethnicity'] == 'black']
chart_black = roster_black['year'].value_counts().reset_index()
chart_black.columns = ['Year','Amount']
chart_black['Year'] = chart_black['Year'].astype(str)

In [60]:
chart_black.style.set_caption("Black Officers from 2010 - 2021")

In [61]:
test = chart_black['Amount'].mean()
test

365.3333333333333

In [62]:
black_chart = alt.Chart(chart_black, title= 'Black Officers from 2010 - 2021').mark_bar(color="#FE2D00").encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

In [63]:
black_chart

alt.Chart(...)

## **Overall, the number of Black officers decreased in the past 10 years. Hit its lowest point in 2020, not counting 2021 since that data could change.**

### Hispanic Officer Change from 2010 - 2021

In [64]:
roster_hispanic = year_race.loc[year_race['clean_race_ethnicity'] == 'hispanic']
chart_hispanic = roster_hispanic['year'].value_counts().reset_index()
chart_hispanic.columns = ['Year','Amount']
chart_hispanic['Year'] = chart_hispanic['Year'].astype(str)

In [65]:
hispanic_chart = alt.Chart(chart_hispanic, title= 'Hispanic Officers from 2010 - 2021').mark_bar().encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

In [66]:
hispanic_chart

alt.Chart(...)

## **Interesting to see that Hispanic officers more than doubled over the past ten years**

### Multiracial Officer Change from 2010 - 2021

In [67]:
roster_multi = year_race.loc[year_race['clean_race_ethnicity'] == 'multiracial']
chart_multi = roster_multi['year'].value_counts().reset_index()
chart_multi.columns = ['Year','Amount']
chart_multi['Year'] = chart_multi['Year'].astype(str)

In [68]:
multi_chart = alt.Chart(chart_multi, title= 'Multiracial Officers from 2010 - 2021').mark_bar().encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

In [69]:
multi_chart

alt.Chart(...)

## **Large amounts of officers, 60+, designated as multiracial starting in 2018. Definitely affect other racial categories.**

### Asian/Pacific Islander Officer Change from 2010 - 2021

In [70]:
roster_asian = year_race.loc[year_race['clean_race_ethnicity'] == 'asian/pacific islander']
chart_asian = roster_asian['year'].value_counts().reset_index()
chart_asian.columns = ['Year','Amount']
chart_asian['Year'] = chart_asian['Year'].astype(str)

In [71]:
asian_chart = alt.Chart(chart_asian, title= 'Asian/Pacific Islander Officers from 2010 - 2021').mark_bar().encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

In [72]:
asian_chart

alt.Chart(...)

## **Number of Asian/Pacific Islanders never rose above 26 in any given year.**

### Native American Officer Change from 2010 - 2021

In [73]:
roster_native = year_race.loc[year_race['clean_race_ethnicity'] == 'native american']
chart_native = roster_native['year'].value_counts().reset_index()
chart_native.columns = ['Year','Amount']
chart_native['Year'] = chart_native['Year'].astype(str)

In [74]:
native_chart = alt.Chart(chart_native, title= 'Native American Officers from 2010 - 2021').mark_bar().encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

In [75]:
native_chart

alt.Chart(...)

## **Number of Native officers never reached 5.**

### Unknown Officer Change from 2010 - 2021

In [76]:
roster_unknown = year_race.loc[year_race['clean_race_ethnicity'] == 'unknown']
chart_unknown = roster_unknown['year'].value_counts().reset_index()
chart_unknown.columns = ['Year','Amount']
chart_unknown['Year'] = chart_unknown['Year'].astype(str)

In [77]:
unknown_chart = alt.Chart(chart_unknown, title= 'Unknown Officers from 2010 - 2021').mark_bar().encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

In [78]:
unknown_chart

alt.Chart(...)

## Compare White and Black Officer change from 2010 - 2021

In [79]:
#Layered chart with Black stats over chart with white stats for comparison.
white_chart + black_chart

alt.LayerChart(...)

In [80]:
roster.columns

Index(['name', 'job_desc', 'dept_desc', 'gender', 'race_ethnicity',
       'date_started', 'year', 'clean_name', 'clean_race_ethnicity'],
      dtype='object')

In [81]:
#Staggered dataset to find number of officers and categorize them by gender and year.
all_year_gender = roster.groupby(['year','gender'])['gender'].count().to_frame(name='amount').reset_index()

In [82]:
all_year_gender

year gender  amount
0   2010      F     466
1   2010      M    1436
2   2011      F     441
3   2011      M    1375
4   2012      F     532
5   2012      M    1497
6   2013      F     457
7   2013      M    1476
8   2014      F     467
9   2014      M    1463
10  2015      F     469
11  2015      M    1485
12  2016      F     473
13  2016      M    1541
14  2017      F     497
15  2017      M    1527
16  2018      F     546
17  2018      M    1559
18  2019      F     479
19  2019      M    1495
20  2020      F     489
21  2020      M    1488
22  2021      F     500
23  2021      M    1487

In [83]:
alt.Chart(all_year_gender).mark_bar().encode(
    alt.X('gender', axis=alt.Axis(labelAngle=0)),
    alt.Y('amount'),column='year').properties(
    width=100,
    height=300
)

alt.Chart(...)

## **Females officers remained below 600 for past 10 years. Highest number was 546 in 2018 and lowest was 441 in 2011**
## **Male officers remained above 1400 officers hitting a peak of 1559 in 2018**

# **Started comparing Male and Female officer makeup over time**

In [84]:
#Pulled year and gender from dataset
year_gender = roster[['year','gender']]

### Started creating chart that shows change in male officers over time

In [85]:
#Filtered data to only include male officers.
roster_male = year_gender.loc[year_gender['gender'] == 'M']
#Found how many male officers there were by year
chart_male = roster_male.value_counts().reset_index()
#No need for gender column anymore since they're all male
del chart_male['gender']
#Renamed columns for clarity
chart_male.columns = ['Year', 'Amount']
#Turned the year column into string, because Altair takes in years as integers and starts chart at 0.
chart_male['Year'] = chart_male['Year'].astype(str)

In [86]:
male_chart = alt.Chart(chart_male).mark_bar().encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

### Notes above are same for following chart
### Below, I am creating a chart that shows the change in female officers over time

In [87]:
roster_female = year_gender.loc[year_gender['gender'] == 'F']
chart_female = roster_female.value_counts().reset_index()
del chart_female['gender']
chart_female.columns = ['Year', 'Amount']
chart_female['Year'] = chart_female['Year'].astype(str)

In [88]:
female_chart = alt.Chart(chart_female).mark_bar(color="#FE2D00").encode(
    alt.X('Year', axis=alt.Axis(labelAngle=0)),
    alt.Y('Amount')).properties(
    width=800,
    height=300
)

In [89]:
#Combined both charts for comparison purposes. Red is female, blue is male.
male_chart + female_chart

alt.LayerChart(...)

### Started seeing total number of officers at MNPD over time

In [90]:
#Categorized data by year and found the count of every year to see how many officers there were per year.
total_amount = roster.groupby('year')['year'].count().to_frame(name='amount').reset_index()
#Turned year into string for Altair purposes since it will start it at 0 if it remains an integer.
total_amount['year'] = total_amount['year'].astype(str)

In [91]:
total_amount

year  amount
0   2010    1902
1   2011    1816
2   2012    2029
3   2013    1933
4   2014    1930
5   2015    1954
6   2016    2014
7   2017    2024
8   2018    2105
9   2019    1974
10  2020    1977
11  2021    1987

In [92]:
#Average number of officers from 2010 to 2021
total_amount['amount'].mean()

1970.4166666666667

In [93]:
alt.Chart(total_amount, title = 'Number of Officers at MNPD Over Time').mark_bar().encode(
    alt.X('year', axis=alt.Axis(labelAngle=0)),
    alt.Y('amount')).properties(
    width=800,
    height=500
)

alt.Chart(...)

## Hovered above 1,800 hitting a peak of 2,105 in 2018.

In [94]:
roster.columns

Index(['name', 'job_desc', 'dept_desc', 'gender', 'race_ethnicity',
       'date_started', 'year', 'clean_name', 'clean_race_ethnicity'],
      dtype='object')

In [95]:
roster.groupby(['year','clean_race_ethnicity','clean_name'])['clean_name'].count().to_frame(name='amount').reset_index().groupby(['year','clean_race_ethnicity'])['clean_name'].count().to_frame().reset_index().groupby(['clean_race_ethnicity'])['clean_name'].mean()

clean_race_ethnicity
asian/pacific islander      23.666667
black                      365.333333
hispanic                    35.250000
multiracial                 64.250000
native american              2.500000
unknown                      1.875000
white                     1520.666667
Name: clean_name, dtype: float64

In [96]:
roster.columns

Index(['name', 'job_desc', 'dept_desc', 'gender', 'race_ethnicity',
       'date_started', 'year', 'clean_name', 'clean_race_ethnicity'],
      dtype='object')

In [97]:
test = roster.groupby(['year','gender'])['clean_name'].count().to_frame().reset_index()
test

year gender  clean_name
0   2010      F         466
1   2010      M        1436
2   2011      F         441
3   2011      M        1375
4   2012      F         532
5   2012      M        1497
6   2013      F         457
7   2013      M        1476
8   2014      F         467
9   2014      M        1463
10  2015      F         469
11  2015      M        1485
12  2016      F         473
13  2016      M        1541
14  2017      F         497
15  2017      M        1527
16  2018      F         546
17  2018      M        1559
18  2019      F         479
19  2019      M        1495
20  2020      F         489
21  2020      M        1488
22  2021      F         500
23  2021      M        1487

In [98]:
test_21 = test.loc[test['year']==2021]

In [99]:
test_21['percentage'] = (test_21['clean_name']/test_21['clean_name'].sum())*100
test_21.round(1)

<ipython-input-99-a6d16c9380e5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_21['percentage'] = (test_21['clean_name']/test_21['clean_name'].sum())*100


year gender  clean_name  percentage
22  2021      F         500        25.2
23  2021      M        1487        74.8

In [100]:
roster

name  \
0      PETTY, MARLENE F                           
1      DUPIE III, LEANDER H.                      
2      TAYLOR, CHRIS M.                           
3      DYER, WILLIAM K.                           
4      SOWELL III, CAMPBELL M.                    
...                                         ...   
23640                   Modransky, Eric Charles   
23641                       Robertson, Joshua E   
23642                Vanders, Christian William   
23643                      Vegas, Parker Joseph   
23644                        Williams, Darryl W   

                             job_desc                       dept_desc gender  \
0      Police Operations Coord 2       Police                              F   
1      Police Lieutenant               Police                              M   
2      Police Captain                  Police                              M   
3      Police Lieutenant               Police                              M   
4      Police Sergeant                 Police                              M   
...                               ...                             ...    ...   
23640                Police Officer 1                      031 Police      M   
23641                Police Officer 2                      031 Police      M   
23642                Police Officer 2                      031 Police      M   
23643                Police Officer 2                      031 Police      M   
23644                Police Officer 2                      031 Police      M   

                       race_ethnicity date_started  year  \
0      White (Not of Hispanic Origin)   1998-02-01  2010   
1      White (Not of Hispanic Origin)   1991-11-01  2010   
2      White (Not of Hispanic Origin)   1992-03-16  2010   
3      White (Not of Hispanic Origin)   1988-03-16  2010   
4      White (Not of Hispanic Origin)   1992-09-01  2010   
...                               ...          ...   ...   
23640                           White   2019-08-16  2021   
23641                           White   2006-12-16  2021   
23642                           White   2019-03-16  2021   
23643                           White   2019-03-16  2021   
23644       Black or African American   2014-04-01  2021   

                      clean_name clean_race_ethnicity  
0                f marlene petty                white  
1            dupie h iii leander                white  
2                 chris m taylor                white  
3                 dyer k william                white  
4          campbell iii m sowell                white  
...                          ...                  ...  
23640     charles eric modransky                white  
23641         e joshua robertson                white  
23642  christian vanders william                white  
23643        joseph parker vegas                white  
23644          darryl w williams                black  

[23645 rows x 9 columns]